pkg/k8s/authz.go

```diff
@@ -105,6 +107,39 @@ func ServiceProfilesAccess(k8sClient kubernetes.Interface) error {
 	return errors.New("ServiceProfile CRD not found")
 }
 
+// EndpointSliceAccess verifies whether the K8s cluster has
+// access to EndpointSlice resources
+func EndpointSliceAccess(k8sClient kubernetes.Interface) error {
+	gv := discovery.SchemeGroupVersion.String()
+	res, err := k8sClient.Discovery().ServerResourcesForGroupVersion(gv)
+	if err != nil {
+		return err
+	}
+
+	if res.GroupVersion == gv {
+		for _, apiRes := range res.APIResources {
+			if apiRes.Kind == "EndpointSlice" {
+				return checkEndpointSlicesExist(k8sClient)
+			}
+		}
+	}
+
+	return errors.New("EndpointSlice resource not found")
+}
+
+func checkEndpointSlicesExist(k8sclient kubernetes.Interface) error {
```

Comment:
I've been thinking a bit more about this and I think it would be a good idea to have the EndpointSlices functionality be opt-in.  This has a few advantages:

* it seems safer to let users opt-in when they know it has been enabled on their cluster rather than us trying to guess based on the existence of an EndpointSlices object
* it allows to us to release this sooner since it is explicitly opt-in and experimental
* we can change the default behavior in the future (perhaps once kube-proxy uses EndpointSlices by default)
* If we discover issues, this gives us a mechanism to fall back to Endpoints mode.

In practice, I see this as being an install flag (and helm variable) that enables EndpointSlices mode.  We should also include some validation that prevents EndpointSlices mode from being enabled on a cluster without the EndpointSlices API resource type.

In the interest of getting this merged and rolling forward, I think the safest thing to do would be to comment out the body of the EndpointSliceAccess function and hard-code it to return false, effectively making all of the EndpointSlice logic unreachable.  Then, in a separate PR, we can add the install flag which enables EndpointsSlice mode.  What do you think?

Reply:
@adleong it makes sense, and I think it's a good decision. I have been doubting the use of the access function lately, especially after I've read that in K8s v1.8 `EndpointSlices` are going to be switched on by default (but not in-use by kube-proxy).

Realistically, as much as I have tried to be thorough with both the unit tests and the manual tests, this needs to be played around with for a while so we can see where unconventional edge cases might be; I'm all for hardcoding the function to return a false and leaving it out for now.

Will make the change and maybe open up an issue in the next few days to get the ball rolling on how to enable them. So far I like the idea of an install flag, can maybe make it part of an edge release and hide the command (like the `alpha stat` cmd perhaps?).